### ISSUE TITLE
[Data] Pipelines create too many Actors

### ISSUE THREAD

**Issue by thatcort** [ray-project/ray#26396]
### What happened + What you expected to happen

When using Actors with a Dataset Pipeline, the pipeline creates a new instance of the Actor per block processed. This ignores the ActorPoolStrategy and results in a lot of instantiation overhead that Actors are meant to avoid.

Here is the output of running the included reproduction script. You can see that the actor is instantiated for every block (6 times) instead of only once as specified in the ActorPoolStrategy.
```
Found local data file
Repartition: 100%|██████████| 6/6 [00:00<00:00,  7.75it/s]
2022-07-08 18:26:11,488	INFO dataset.py:2975 -- Created DatasetPipeline with 3 windows: 0.01MiB min, 0.02MiB max, 0.02MiB mean
Stage 0:   0%|          | 0/3 [00:00<?, ?it/s]
  0%|          | 0/3 [00:00<?, ?it/s]
Stage 1:   0%|          | 0/3 [00:00<?, ?it/s](BlockWorker pid=4301) Actor init...
(BlockWorker pid=4301) Actor init... Done.
(BlockWorker pid=4301) Processing...
(BlockWorker pid=4301) Processing... Done
(BlockWorker pid=4301) Processing...

Stage 0:  67%|██████▋   | 2/3 [00:02<00:01,  1.36s/it](BlockWorker pid=4301) Processing... Done
Finished batch.
Finished batch.
(BlockWorker pid=4336) Actor init...
(BlockWorker pid=4336) Actor init... Done.
(BlockWorker pid=4336) Processing...
(BlockWorker pid=4336) Processing... Done
(BlockWorker pid=4336) Processing...

Stage 0: 100%|██████████| 3/3 [00:05<00:00,  1.92s/it]Finished batch.
Finished batch.
(BlockWorker pid=4336) Processing... Done
(BlockWorker pid=4370) Actor init...
(BlockWorker pid=4370) Actor init... Done.
(BlockWorker pid=4370) Processing...
(BlockWorker pid=4370) Processing... Done
(BlockWorker pid=4370) Processing...

Stage 1: 100%|██████████| 3/3 [00:08<00:00,  2.71s/it]
Stage 0: 100%|██████████| 3/3 [00:08<00:00,  2.72s/it]Finished batch.
Finished batch.
```

### Versions / Dependencies

This is using Ray 1.13.0 and Pyarrow 6.0.1 on Python 3.7.13 running on Google Colab.

### Reproduction script

```py
import time
class Noop:
  def __init__(self):
    print('Actor init...')
    time.sleep(0.5)
    print('Actor init... Done.')

  def __call__(self, batch):
    print('Processing...')
    time.sleep(0.5)
    print('Processing... Done')
    return batch

table = loadData(data_name)
table = table.slice(length=200)
if not ray.is_initialized():
  ray.init()
ds = ray.data.from_arrow(table)
pipe = ds.repartition(num_blocks=6).window(blocks_per_window=2)
pipe = pipe.map_batches(Noop, compute=ray.data.ActorPoolStrategy(1, 1), batch_format="pyarrow", batch_size=None)
for b in pipe.iter_batches():
  print('Finished batch.')
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by clarkzinzow** [comment]
Hey @thatcort, thank you for filing the issue! I can reproduce this on latest master, this looks like an issue with sharing the actor pool strategy across windows of a pipeline.

**Comment by ericl** [comment]
This is an interesting issue: the reason the actors were re-created each window is that this avoids potential complications involving deadlock. Imagine for example two stages that both require GPU actors, but there is only 1 GPU available. Without this logic, deadlock could happen.

However, recreating actors does add overhead especially if windows are small. We could consider adding a mode to retain the actor pools (and require users manually specify the sizes properly to avoid deadlocks).

**Comment by thatcort** [comment]
That's an interesting point. Since Ray knows both the resources available and required by every stage in the pipeline, can't it at least determine a minimum number of actors of each type that it can keep around. Then the rest can be scaled based on backpressure?

**Comment by ericl** [comment]
Yup. You can imagine an alternative execution strategy for DatasetPipelines would be to create a worker pool (if using actors) for each stage, and autoscaling would operate at the entire pipeline level, scaling the size of each worker pool up/down. This would be an alternative / potential replacement for the current strategy of leveraging the Datasets for execution only without persistent state.

I think we should seriously consider this approach in the future (afaik it's a common strategy for pipelined execution engines), cc @jianoaix and @c21 for thoughts.

**Comment by dirtyValera** [comment]
Hi @ericl, @clarkzinzow is there any update on this? Or a workaround?

I'm planning to run a transform pipe using Dataset pipelines + ActorPool in conjunction with Kuberay Autoscaler to have elastic execution bound to specific resource quota range. If Ray keeps spawning Actors for this set up I see no way to bound pipeline's resource consumption (only way is to limit number of nodes for autoscaler based on max number of actors for specific actor type (using custom resources?), which looks kind of ugly and I'm not sure will work).

@thatcort, have you managed to bypass this?

Any suggestion is welcome, thanks!

**Comment by ericl** [comment]
@dirtyValera have you tried the max actors setting for the ActorPoolStrategy? It's new in 2.0: https://docs.ray.io/en/master/data/transforming-datasets.html#compute-strategy

I think with max actors set, the resources will be bounded since the old actors are destroyed before the new ones are launched.

In the longer term, we are planning to improve Datasets' pipelining support to use the same pool of actors for the entire computation--- look for a REP on this soon.

**Comment by thatcort** [comment]
@dirtyValera I don't have a workaround. I'm avoiding using pipelines because of this bug.

**Comment by percevalw** [comment]
Hi, I have the same problem. Using a pipeline instead of processing all the data stage by stage leads to a significant increase (x 2 or more) in processing time, since loading my model into an actor takes between 20 and 30 seconds, and pipelining is necessary for large datasets in my case (reading from hdfs parquets or s3 files → processing → writing to hdfs parquet files).
Unfortunately, setting up an ActorPoolStrategy did not improve performance as the actor seem to be destroyed and created after processing each window.

**Comment by ericl** [comment]
Update here: We have a REP for this issue https://github.com/ray-project/enhancements/pull/18

Optimistically, this may be available as an experimental feature in 2.3.

**Comment by ericl** [comment]
Happy to announce this is available as a dev preview feature in 2.3! Here's a guide for how to try out the new streaming execution mode: https://docs.google.com/document/d/1BXd1cGexDnqHAIVoxTnV3BV0sklO9UXqPwSdHukExhY/edit#heading=h.dkdm0ffungpx

The plan is to make this the default execution backend for Dataset is 2.4, and gradually deprecate the DatasetPipeline functionality.
